### PR TITLE
feat: ensure images are processed by configurable sample ids

### PR DESCRIPTION
**Pull request by neerajaabhyankar**
## 🔗 Related Issues

Internal context for issue: https://voxel51.slack.com/archives/C08KF01388P/p1777724100560819?thread_ts=1776433614.028989&cid=C08KF01388P

## 📋 What changes are proposed in this pull request?

Previously, the `index_to_instance` method operated on samples with the same `id` at a time, since instance IDs only intend to be shared within the same video frame. This did not work for frames views of the video or dynamically grouped datasets, since frames of the video are most commonly identified by the same `sample_id`.

This PR makes the ID field (among which to track the same instances) configurable.

## 🧪 How is this patch tested? If it is not, please explain why.

Added two new unit tests that reproduced the issue, that pass with these changes
- `test_label_to_instance_frames_view`
- `test_label_to_instance_frames_view_grouped`

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

The `fiftyone.utils.labels.index_to_instance` method now also works on `to_frames` views of a video, as well as on dynamically grouped datasets.

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

* **Bug Fixes**
  * More reliable label-to-instance mapping for dynamic groups and grouped/frame collections; added support for specifying a custom ID field, safer fallback when ID paths are missing, and per-group/frame updates that only modify the selected view (optionally clearing original index values).

* **Tests**
  * Added a shared helper and new tests covering frames views and grouped frames ordering to validate instance/index distributions; minor test formatting tweaks.

[Review Change Stack](https://app.coderabbit.ai/change-stack/voxel51/fiftyone/pull/7496?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: fiftyone/utils/labels.py [hunk 1]
```diff
@@ -950,15 +959,18 @@ def index_to_instance(
...
-
-        sample_collection = sample_collection.select_group_slices(
-            _allow_mixed=True
-        )
+        if sample_collection._is_dynamic_groups:
+            id_path = id_field
+            ids = list(dict.fromkeys(sample_collection.values(id_path)))
+        else:
+            id_path = sample_collection.group_field + "." + id_field
+            ids = list(dict.fromkeys(sample_collection.values(id_path)))
```
Comment (benjaminpkane) [inline fiftyone/utils/labels.py]: Out of curiosity, is the below not sufficient?
```suggestion
            ids = list(sample_collection.values(id_path))
```

File: fiftyone/utils/labels.py [hunk 2]
```diff
@@ -923,6 +924,14 @@ def index_to_instance(
             :class:`fiftyone.core.labels.Polylines`,
             :class:`fiftyone.core.labels.Keypoint`, and
             :class:`fiftyone.core.labels.Keypoints`
+        id_field ("id"): the attribute identifying samples/frames of the
+            same video/sequence. Samples with the same id_field are processed
+            together. For dynamic grouped views from
+            :meth:`fiftyone.core.collections.SampleCollection.group_by` with
+            ``flat=False``, this must be the group key (e.g. ``"sample_id"`` when
+            grouping frames by parent video). For pinned group datasets,
```
Comment (benjaminpkane) [inline fiftyone/utils/labels.py]: What is a pinned group dataset? A group dataset with slices? e.g. left/right/pcd in quickstart-groups?

File: fiftyone/utils/labels.py [hunk 3]
```diff
@@ -923,6 +924,14 @@ def index_to_instance(
...
+            ``id_field`` names an attribute under the dataset's ``group_field``
+            (the default ``"id"`` refers to ``<group_field>.id``)
```
Comment (benjaminpkane) [inline fiftyone/utils/labels.py]: Based on my knowledge of group datasets (and dynamically grouped views of datasets) this language might be more consistent. Assuming pinned group datasets are group datasets, e.g. quickstart-groups
```suggestion
        id_field ("id"): the attribute identifying samples of the same
            sequence, e.g. image frames of a dynamically grouped video. Samples
            with the same ``id_field`` are processed together. For dynamically
            grouped views created via
            :meth:`fiftyone.core.collections.SampleCollection.group_by` with
            ``flat=False``, this must be the group key (e.g. ``"sample_id"``
            when grouping frames for a video). For group datasets, ``id_field``
            names an attribute under the dataset's ``group_field`` (the default
            ``"id"`` refers to ``<group_field>.id``)
```

**Review by benjaminpkane — COMMENTED** [review]
I'll defer to ML for final review

File: fiftyone/utils/labels.py (hunk 3 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: > A group dataset with slices? e.g. left/right/pcd in quickstart-groups?

Yes, that's what I meant. I'll use your language, seems more standard, thanks!

File: fiftyone/utils/labels.py (hunk 1 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: It fails for the `image` view or when we do indeed have multiple samples with the same id_path:
`ids` should be a deduplicated list, but list() returns the full list of len(sample_collection) and ends up processing each separately `for id in pb(ids)`

**Review by neerajaabhyankar — COMMENTED** [review]
Thanks for the review!

File: fiftyone/utils/labels.py (hunk 1 above)
Comment (benjaminpkane) [reply fiftyone/utils/labels.py]: Ok, let me think about this. I can review again tomorrow AM

File: fiftyone/utils/labels.py [hunk 4]
```diff
@@ -949,16 +961,29 @@ def index_to_instance(
...
+        not sample_collection._is_dynamic_groups
+    ):
+        id_path = sample_collection.group_field + "." + id_field
+        ids = list(dict.fromkeys(sample_collection.values(id_path)))
+        if None in ids:
+            logger.warning(
+                "Some samples have no %s; defaulting to 'id'", id_path
+            )
+            id_path = sample_collection.group_field + "." + "id"
+            ids = list(dict.fromkeys(sample_collection.values(id_path)))
```
Comment (benjaminpkane) [inline fiftyone/utils/labels.py]: Can we add a test case that leverages a custom `id_field` on a group dataset?

And also a test case that asserts fallback when `None` is observed 😄

Comment (benjaminpkane) [reply fiftyone/utils/labels.py]: Have we also considered the case of dynamically grouped "group datasets". For example, there is [NuScenes](https://demo.fiftyone.ai/datasets/nuscenes/samples?view=scenes). Each timestamp has multiple slices.

File: fiftyone/utils/labels.py [hunk 5]
```diff
@@ -949,16 +961,29 @@ def index_to_instance(
...
-        id_path = "id"
-        ids = sample_collection.values(id_path)
+        id_path = id_field
+        ids = list(dict.fromkeys(sample_collection.values(id_path)))
+        if None in ids:
+            logger.warning(
+                "Some samples have no %s; defaulting to 'id'", id_path
+            )
+            id_path = "id"
+            ids = list(dict.fromkeys(sample_collection.values(id_path)))
```
Comment (benjaminpkane) [inline fiftyone/utils/labels.py]: I think this code path is missing from the new tests too

File: fiftyone/utils/labels.py [hunk 6]
```diff
@@ -923,6 +924,17 @@ def index_to_instance(
             :class:`fiftyone.core.labels.Polylines`,
             :class:`fiftyone.core.labels.Keypoint`, and
             :class:`fiftyone.core.labels.Keypoints`
+        id_field ("id"): the attribute identifying samples of the same
+            sequence, e.g. image frames of a dynamically grouped video. Samples
+            with the same ``id_field`` are processed together. For dynamically
+            grouped views created via
+            :meth:`fiftyone.core.collections.SampleCollection.group_by` with
+            ``flat=False``, this must be the group key (e.g. ``"sample_id"``
```
Comment (benjaminpkane) [inline fiftyone/utils/labels.py]: Should we assert this requirement? I don't see an exception raised when this is not the case. Or do we need to note this? Note that a dynamic group can also be an expression.

I'm not sure why we would want to support `expr` for this utility, but it is a valid view.
```py
from fiftyone import ViewField as F
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart")

# Group samples by the number of ground truth objects they contain
expr = F("ground_truth.detections").length()
view = dataset.group_by(expr)
```

**Review by benjaminpkane — CHANGES_REQUESTED** [review]
These are definitely needed changes. Can we add test cases for some of the other decisions made 😄 

I see we handle non-dynamic group datasets differently now.

Tests that handle the presence of `None`, the fallback decision, and behavior around dynamically grouped group datasets is important I think

File: fiftyone/utils/labels.py (hunk 6 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: > Note that a dynamic group can also be an expression.

Ahh okay. Then I'll get rid of the "this must be the group key", has the assumption that we only group by video sequences.
Will add some tests and edit this.

File: fiftyone/utils/labels.py (hunk 4 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: > dynamically grouped groups

handled with a bunch of [if statements](https://github.com/voxel51/fiftyone/blob/c4383251c99e56b7f6cec8aeebe9b1dc8379f2d2/fiftyone/utils/labels.py#L978) (couldn't figure out anything cleaner), let me know if you think there could be a better way!

File: fiftyone/utils/labels.py (hunk 5 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: added this to `test_label_to_instance_dynamic_grouped_group`

File: fiftyone/utils/labels.py (hunk 4 above)
Comment (benjaminpkane) [reply fiftyone/utils/labels.py]: Looks great!

**Review by benjaminpkane — APPROVED** [review]
Pardon the delay, and thanks for the adjustments. Looks great! 🚀  Groupings can be a rabbit hole 🐇

File: fiftyone/utils/labels.py [hunk 7]
```diff
@@ -949,16 +957,42 @@ def index_to_instance(
         ),
     )
 
-    if sample_collection.media_type == fom.GROUP:
-        id_path = sample_collection.group_field + ".id"
-        ids = sample_collection.values(id_path)
+    is_native_group = (
```
Comment (brimoor) [inline fiftyone/utils/labels.py]: If the user passes in a frames view, should we automatically set id_field="sample_id" so that instance IDs will be reused across all frames associated with a given video sample?

As a user, this is what I'd expect/want to happen if I passed a frames view to this method.

File: fiftyone/utils/labels.py [hunk 8]
```diff
@@ -905,6 +905,7 @@ def classifications_to_detections(
 def index_to_instance(
     sample_collection,
     label_field,
+    id_field="id",
```
Comment (brimoor) [inline fiftyone/utils/labels.py]: I believe these are equivalent, right?

```py
index_to_instance(view, label_field, id_field=id_field)
index_to_instance(view.group_by(id_field), label_field)
```

which would mean that we technically don't need to expose `id_field` as a parameter at all.

I'm curious what the initial motivation for adding the `id_field` parameter was in the first place.

File: fiftyone/utils/labels.py [hunk 9]
```diff
@@ -923,6 +924,13 @@ def index_to_instance(
             :class:`fiftyone.core.labels.Polylines`,
             :class:`fiftyone.core.labels.Keypoint`, and
             :class:`fiftyone.core.labels.Keypoints`
+        id_field ("id"): the attribute identifying samples of the same
```
Comment (brimoor) [inline fiftyone/utils/labels.py]: I would recommend elevating the description you added to the `id_field` parameter to the top-level.

Specifically, the key thing to get across IMO is that this method will reuse instance IDs for all samples within the same group or dynamic group. And that is true even if you don't manually set the `id_field` parameter.

File: fiftyone/utils/labels.py [hunk 10]
```diff
@@ -970,9 +1004,33 @@ def index_to_instance(
...
+            return id_path == group_expr.lstrip("$")
+        if isinstance(group_expr, (list, tuple)):
+            return id_path in (expr.lstrip("$") for expr in group_expr)
+        return False
+
     with fou.ProgressBar(progress=progress) as pb:
-        for id in pb(ids):
-            view = sample_collection.select_by(id_path, id)
+        for uid in pb(ids):
+            if (
```
Comment (brimoor) [inline fiftyone/utils/labels.py]: There's no need to evaluate this `if` statement at every iteration of the loop, right?

File: fiftyone/utils/labels.py [hunk 11]
```diff
@@ -970,9 +1004,33 @@ def index_to_instance(
     else:
         index_paths = [index_path]
 
+    def _id_path_matches_dynamic_group_by_key(sample_collection, id_path):
```
Comment (brimoor) [inline fiftyone/utils/labels.py]: Oh actually what I said above is not true. This implementation does not automatically reuse sample IDs within dynamic groups. 

You could easily make it so, however. For dynamic grouped views whose key is a simple string, you could just set `id_path` to the dynamic group's group by key.

For other dynamic grouped views, the grouping is already done for you! Just iterate over `iter_dyanmic_groups()` rather than using the `for loop` + `select_by()` pattern.

Comment (brimoor) [reply fiftyone/utils/labels.py]: I think it would be cleaner if we didn't expose `id_field` to the user and instead required them to pass in a grouped view if they wanted to reuse instance IDs between groups of samples.

In this case, it would be impossible for a sample to have a missing group:
- **Dynamic group views:** by definition, every sample belongs to a group
- **Frame views:** `sample_id` is always present on all samples
- **Everything else:** `id` is required on all samples

**Review by brimoor — CHANGES_REQUESTED** [review]

File: fiftyone/utils/labels.py (hunk 8 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: They'll be equivalent if we infer the id_field from the group_by clause, but not if we allow the flexibility of, say, grouping by multiple keys but identifying samples by only one of them.

File: fiftyone/utils/labels.py (hunk 7 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: Defaulting to `sample_id` instead of `id` sounds good.

Although, in the case of a native image dataset imported as individual frames (and not as a video followed by `to_frames`), the id_field can be arbitrary. (e.g., [sequence_id for the davis-2017 dataset](https://github.com/voxel51/davis-2017#sample-fields))

File: fiftyone/utils/labels.py (hunk 11 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: I think this is a good idea.
> Just iterate over iter_dyanmic_groups() rather than using the for loop + select_by() pattern

As long as users wanting to re-use only within the given groups is a valid assumption.

> required them to pass in a grouped view

And as long as this is clear in the docstring as the favored way of handling an image dataset with an arbitrarily-named sample_id field.

Will do a refactor.

**Review by neerajaabhyankar — COMMENTED** [review]
Going with not exposing `id_field` and inferring it appropriately. Will push a commit soon.

File: fiftyone/utils/labels.py (hunk 10 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/labels.py]: post refactor: still left this inside the loop
```
                if (
                    view.media_type == fom.GROUP
                    and view.group_field is not None
                    and not view._is_dynamic_groups
                ):
```
but it's an O(1) evaluation; need to use this flow for `view = view.select_group_slices(_allow_mixed=True)` at each iteration anyway

**Review by neerajaabhyankar — COMMENTED** [review]
Got rid of the `id_field` attribute -- happy to edit the docstring if the resolution mechanism is unclear :)
